### PR TITLE
Switch elliptic executable to generic observers

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

This PR removes the dependence on time for the generic observers so they can be used by the elliptic code. Instead of computing the analytic solution themselves, they retrieve it from the DataBox, where evolution and elliptic code have different ways of placing it.

This is the correct order of commits (from `develop` to `HEAD`):

- <s>Expose floating-point time in DataBox</s>
- Add prefix tag for analytic solutions
- Wrap evolution option tags in evolution namespace
- Add compute item for evolution analytic solution
- Add initializer for elliptic analytic solution
- Move generic observe events to ParallelAlgorithms
- Move EventsAndTriggers to ParallelAlgorithms
- Remove dependence on Time from observe events
- Add elliptic Triggers
- Switch elliptic executable to generic observers

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by nilsvu** [comment]
@wthrowe To make this PR easier, the first commit "Expose floating-point time in DataBox" does a subset of the changes from your #1529. We can get that PR in soon and I rebase this one on top of it, if you prefer.

**Comment by nilsvu** [comment]
rebased on merged #1529. @wthrowe could you review this?

**Comment by wthrowe** [comment]
Doesn't compile.

**Comment by wthrowe** [comment]
(You may need the commit "Update some Tag::name()..." from #1681, but there are other problems unrelated to that as well.)

**Comment by nilsvu** [comment]
@wthrowe there was only a left-over `#include "Time/TimeId.hpp"`. Should compile and pass tests now :)

File: src/ParallelAlgorithms/Events/ObserveErrorNorms.hpp
Comment (wthrowe) [inline src/ParallelAlgorithms/Events/ObserveErrorNorms.hpp]: Since the non-type template parameters are gone, this can be just `using ObserveErrorNorms = Registration::Registrar<Events::ObserveErrorNorms, TemporalId, Tensors>;`

File: src/Elliptic/Triggers/EveryNIterations.hpp
Comment (wthrowe) [inline src/Elliptic/Triggers/EveryNIterations.hpp]: Use `Registration::Registrar`.

**Review by wthrowe — CHANGES_REQUESTED** [review]
Just a minor thing.  Squash.

**Comment by nilsvu** [comment]
squashed this in, waiting for Travis now

**Comment by nilsvu** [comment]
Had to rebase on develop (for the `const_global_cache_tags` changes). Travis passed on my fork (https://travis-ci.org/nilsleiffischer/spectre/builds/584727663). @kidder @nilsdeppe could you provide the second review or suggest a reviewer?

**Comment by nilsdeppe** [comment]
could you please factor out the code that is more generic than your PR implies? Your PRs always have a lot of collateral changes.

**Comment by nilsvu** [comment]
@wthrowe you reviewed this already. Do you think this would benefit from being split into multiple PRs?

@nilsdeppe please see #1689. Do you have a particular change you would like me to factor out?

**Comment by nilsdeppe** [comment]
@kidder basically summarized our discussion well on #1689. I think this PR should be split into the following:

PR A:
- Add prefix tag for analytic solutions
- Wrap evolution option tags in evolution namespace
- Add compute item for evolution analytic solution

PR B:
- Add initializer for elliptic analytic solution

PR C:
- Move generic observe events to ParallelAlgorithms

PR D:
- Move EventsAndTriggers to ParallelAlgorithms

PR E:
- Remove dependence on Time from observe events

PR F:
- Add elliptic Triggers
- Switch elliptic executable to generic observers

This might seem like a lot, but this way each PR does one thing. As it stands, this PR does way too many things in different places. To be clear, it is not the _number of lines_ that I dislike, we want people to write good tests, and those alone more than double the amount of code in a commit, but that there are a large number of completely unrelated changes in this PR. The name of the PR suggests something along the lines of PR F, definitely not all the other changes. For example, I would like to be able to have an opinion about PR A and PR C, but the others I don't feel strongly about (of course I'm happy to give opinions/advice/reviews if asked). By having so many unrelated changes in a single PR it feels like an attempt to "sneak" code in without making it easy for the appropriate people to easily identify a PR as something they may care about. I realize this is not being done intentionally. A lot of companies use diffs instead of PRs, which would translate here to each PR being only 1 commit. I don't think this is a model we need to pursue as long as PRs are single logical units. While all the changes here are necessary for the elliptic code to perform observations using the generic code, that doesn't make the changes related. For example, I could submit a 30k line PR with the title `Improve disk evolutions` where it really adds subcell, modifies the time steppers, changes the way initial data and boundary conditions are handled, and completely rewrites the way we compute time derivatives. The description isn't wrong, it's just that the PR is doing way more than is eluded to by the description. That's an extreme case of what is being done in this PR.

**Comment by nilsvu** [comment]
@nilsdeppe I agree with a lot of what you say, thanks for spelling out your arguments :) I dread the >72 hours that Travis will probably work on these 6 PRs, but here's the first batch: #1690 #1691 #1692 #1693

**Comment by nilsdeppe** [comment]
Yes, I agree that the travis runtime will be terrible.. But it will also be a good test as to whether fresh PRs actually properly use the develop cache :) Thank you for splitting this up :)

**Comment by kidder** [comment]
@nilsleiffischer can this be closed?

**Comment by nilsvu** [comment]
@kidder No, I'll update this PR once #1703 is merged.

**Comment by nilsvu** [comment]
@wthrowe @nilsdeppe @kidder with #1703 merged this PR has now arrived at (the final) stage "F" of @nilsdeppe's comment above. It only affects elliptic code. Please review or assign reviewers :)

File: src/Elliptic/Triggers/EveryNIterations.hpp
Comment (wthrowe) [inline src/Elliptic/Triggers/EveryNIterations.hpp]: `db::const_item_type` due to recent changes elsewhere.

File: src/Elliptic/Triggers/EveryNIterations.hpp
Comment (wthrowe) [inline src/Elliptic/Triggers/EveryNIterations.hpp:10]: `#include "DataStructures/DataBox/DataBoxTag.hpp"`

File: tests/Unit/Elliptic/Triggers/Test_EveryNIterations.cpp
Comment (wthrowe) [inline tests/Unit/Elliptic/Triggers/Test_EveryNIterations.cpp:21]:
```
#include <cstddef>
#include <initializer_list>
#include <string>
#include "DataStructures/DataBox/DataBoxTag.hpp"
```
(IWYU also wants `<vector>` and `"Utilities/Registration.hpp"`, but I don't see any uses of those.)

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: tests/Unit/Elliptic/Triggers/Test_EveryNIterations.cpp
Comment (wthrowe) [inline tests/Unit/Elliptic/Triggers/Test_EveryNIterations.cpp]: And now you don't need `<string>` anymore.

**Review by wthrowe — COMMENTED** [review]
No one else has reviewed so far, so go ahead and squash.

File: tests/Unit/Elliptic/Triggers/Test_EveryNIterations.cpp
Comment (nilsvu) [reply tests/Unit/Elliptic/Triggers/Test_EveryNIterations.cpp]: is it not needed for the call to `test_factory_creation`?

Comment (wthrowe) [reply tests/Unit/Elliptic/Triggers/Test_EveryNIterations.cpp]: You're right, it is.  IWYU is confused, I guess.

**Review by wthrowe — DISMISSED** [review]

**Comment by nilsvu** [comment]
@sxs-collaboration/spectre-core-devs Could you provide the second review or assign a reviewer?

**Review by nilsdeppe — DISMISSED** [review]

File: src/Elliptic/Triggers/EveryNIterations.hpp
Comment (nilsdeppe) [inline src/Elliptic/Triggers/EveryNIterations.hpp:19]: Since you have to rebase anyway, can you `cond` this?

**Comment by kidder** [comment]
@nilsdeppe @wthrowe please look at

**Review by wthrowe — APPROVED** [review]

**Comment by nilsvu** [comment]
@nilsdeppe please take a look. Thanks for reviewing @nilsdeppe and @nilsdeppe :)

Edit: also thanks for reviewing @wthrowe ;)

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsdeppe** [comment]
Travis failure is GCC6 debug final build

**Pull request merged**
